datacube/index/memory/_datasets.py

```diff
@@ -417,20 +426,23 @@ def archive_location(self, id_: DSID, uri: str) -> bool:
     )
     def restore_location(self, id_: DSID, uri: str) -> bool:
         uuid = dsid_to_uuid(id_)
-        if uuid not in self.archived_locations:
+        if uuid not in self._archived_locations:
             return False
-        old_locations = self.archived_locations[uuid]
+        old_locations = self._archived_locations[uuid]
         new_locations = [(loc, dt) for loc, dt in old_locations if loc != uri]
         if len(new_locations) == len(old_locations):
             return False
-        self.archived_locations[uuid] = new_locations
-        self.locations[uuid].append(uri)
+        self._archived_locations[uuid] = new_locations
+        self._locations[uuid].append(uri)
         return True
 
-    def search_by_metadata(self, metadata: Mapping[str, QueryField]):
-        for ds in self.active_by_id.values():
+    def search_by_metadata(self, metadata: Mapping[str, QueryField], archived: bool | None = False):
+        for ds in self._active_by_id.values():
             if metadata_subset(metadata, ds.metadata_doc):
-                yield ds
+                if archived is None:
+                    yield ds
+                elif archived and ds.is_archived or not (archived or ds.is_archived):
```

Comment:
I feel like this could benefit from a comment

Reply:
Actually, it's broken and needs to be fixed.